### PR TITLE
node-style import

### PR DESCRIPTION
**Pull request by Qard**
This is the beginnings of a node-style import system I'm working on. At this stage, it's usable enough simply supporting a wren_modules folder in the app root. I'd like to get relative import supported also though, for fully node-like nested dependency resolution. See: #325.

I'm hoping that, if I can get nested dependency resolution working, I can build a proper package manager for wren.

### CODE REVIEW

**Comment by munificent** [comment]
Ah, very interesting! Having more powerful import resolution so that we can eventually do a package manager is definitely on the radar. Thanks for hacking on this!

However, I'm not personally a fan of node's nested dependency approach. I wrote the [package manager for Dart](https://pub.dartlang.org/) and it's based on Bundler's model, which [I think is a much better way to do things](https://www.dartlang.org/tools/pub/versioning.html). Even the npm folks seem to be realizing that [nested dependencies aren't all they're cracked up to be](https://nodejs.org/en/blog/npm/peer-dependencies/).

For lots of reasons, I think a flattened dependency graph works better, though it also has its own challenges around constraint resolution. Most package managers I've seen seem to either use Bundler's model or be going in the direction of doing so.

The interesting question is how much support do we need to bake into the Wren CLI in order to support something like Bundler. Should it just walk up directories looking for a lockfile that it can parse? What do you think?

**Comment by Qard** [comment]
I agree that nested dependencies are complicated, and peer dependencies was definitely a mistake.

Personally, I like the current model of npm, supporting nested dependencies, but trying to flatten as much as possible. I'm good with whatever people are comfortable with though. :smile_cat:

Another thing I want to consider for package management is how to support native code. The mruby way of bundling the modules directly into the build is probably better for supporting embedded environments than the node way of loading dynamic libraries. That'd probably add more complexity to supporting nested packages.

File: src/cli/vm.c [hunk 1]
```diff
@@ -116,6 +113,33 @@ static char* readModule(WrenVM* vm, const char* module)
...
+// module was found but could not be read.
+static char* readModule(WrenVM* vm, const char* module)
+{
+  char* source = readBuiltInModule(module);
+  if (source != NULL) return source;
+  
+  source = loadModule(vm, module);
+  if (source != NULL) return source;
+  
+  // If searching in the current path fails, look
```
Comment (paulsapps) [inline src/cli/vm.c]: in wren_modules/ ?

File: src/cli/vm.c [hunk 2]
```diff
@@ -116,6 +113,33 @@ static char* readModule(WrenVM* vm, const char* module)
...
+  char* source = readBuiltInModule(module);
+  if (source != NULL) return source;
+  
+  source = loadModule(vm, module);
+  if (source != NULL) return source;
+  
+  // If searching in the current path fails, look 
+  size_t moduleLength = strlen(module);
+  size_t extModuleLength = moduleLength + 13;
+  char* extModule = (char*)malloc(extModuleLength + 1);
```
Comment (paulsapps) [inline src/cli/vm.c]: where does this get freed?

Comment (Qard) [inline src/cli/vm.c]: Umm...nowhere? 😨

File: src/cli/vm.c (hunk 1 above)
Comment (Qard) [inline src/cli/vm.c]: Yes. I guess I accidentally the rest of the comment. 😉

**Comment by underscorediscovery** [comment]
This PR appears to be out of phase with the latest Wren development, so I'm gonna close this. Discussions about the Wren CLI will soon be on their own repo, where we can pick up the discussion over there.

**Pull request closed without merging**